tests/unit/test_certificates.py

```diff
@@ -0,0 +1,226 @@
+            x509.NameAttribute(x509.NameOID.COMMON_NAME, subject),
+        ]
+    )
+    assert cert.subject == x509.Name(
+        [
+            x509.NameAttribute(x509.NameOID.COUNTRY_NAME, "US"),
+            x509.NameAttribute(x509.NameOID.COMMON_NAME, subject),
+        ]
+    )
+    assert certificate_public_key == initial_public_key
+    assert (
+        x509.KeyUsage(
+            digital_signature=True,
+            key_encipherment=True,
+            key_cert_sign=True,
+            key_agreement=False,
+            content_commitment=False,
+            data_encipherment=False,
+            crl_sign=False,
+            encipher_only=False,
+            decipher_only=False,
+        )
+        == cert.extensions.get_extension_for_class(x509.KeyUsage).value
+    )
+    assert cert.extensions.get_extension_for_class(x509.KeyUsage).critical
+
+
+def test_given_csr_and_ca_when_generate_certificate_then_certificate_is_generated_with_correct_subject_and_issuer():  # noqa: E501
+    ca_subject = "whatever.ca.subject"
+    csr_subject = "whatever.csr.subject"
+    ca_key = generate_private_key_helper()
+    ca = generate_ca_helper(private_key=ca_key, common_name=ca_subject)
+    csr_private_key = generate_private_key_helper()
+    csr = generate_csr_helper(
+        private_key=csr_private_key,
+        common_name=csr_subject,
+    )
+
+    certificate = generate_certificate(
+        csr=csr,
+        ca=ca,
+        ca_key=ca_key,
+    )
+
+    certificate_object = x509.load_pem_x509_certificate(certificate.encode())
+    assert certificate_object.issuer == x509.Name(
+        [
+            x509.NameAttribute(x509.NameOID.COMMON_NAME, ca_subject),
+        ]
+    )
+    subject_name_attributes = certificate_object.subject.get_attributes_for_oid(
+        x509.NameOID.COMMON_NAME
+    )
+    assert subject_name_attributes[0] == x509.NameAttribute(x509.NameOID.COMMON_NAME, csr_subject)
+
+
+def test_given_csr_and_ca_when_generate_certificate_then_certificate_is_generated_with_correct_sans():  # noqa: E501
+    ca_subject = "ca.subject"
+    csr_subject = "csr.subject"
+    sans = ["www.localhost.com", "www.test.com"]
+
+    ca_key = generate_private_key_helper()
+    ca = generate_ca_helper(
+        private_key=ca_key,
+        common_name=ca_subject,
+    )
+    csr_private_key = generate_private_key_helper()
+    csr = generate_csr_helper(
+        private_key=csr_private_key,
+        common_name=csr_subject,
+        sans_dns=sans,
+    )
+
+    certificate = generate_certificate(csr=csr, ca=ca, ca_key=ca_key)
+
+    cert = x509.load_pem_x509_certificate(certificate.encode())
+    result_all_sans = cert.extensions.get_extension_for_class(x509.SubjectAlternativeName)
+
+    result_sans_dns = sorted(result_all_sans.value.get_values_for_type(x509.DNSName))
+    assert result_sans_dns == sorted(set(sans))
+
+
+def test_given_basic_constraint_is_false_when_generate_ca_then_extensions_are_correctly_populated():  # noqa: E501
```

Comment:
What does it mean for the basic constraint to be false?

Reply:
I improved the test name, ultimately it's more that when we generate a certificate using generate_ca, we expect the BasicConstraint extension to be set